lib/id3c/api/datastore.py

```diff
@@ -178,6 +178,39 @@ def store_fhir(session: DatabaseSession, document: str) -> None:
         except (DataError, IntegrityError) as error:
             raise BadRequestDatabaseError(error) from None
 
+@export
+@catch_permission_denied
+def verify_barcode_use_list(session: DatabaseSession, barcode_use_list: str) -> dict:
+    """
+    Check the given *barcode_use_list* containing objects with ``barcode`` and ``use``
+    keys and values to verify that each barcode exists in the backing database and that the
+    given use matches the stored use.
+
+    Returns a list of objects in the same order as the input, with each object including the
+    ``barcode`` (string) being verified, ``barcode_found`` (boolean) indicating whether the
+    given barcode exists, and ``use_match`` (boolean) indicating whether the given use matches
+    the stored use. The ``use_match`` value will be `null` if the barcode does not exist.
+    """
+    barcodes = [bu['barcode'] for bu in barcode_use_list]
+    
+    barcode_use_records = session.fetch_all("""
+        select identifier.barcode, identifier_set.use from warehouse.identifier 
+        join warehouse.identifier_set using (identifier_set_id)
+        where barcode = ANY(%s)""", (barcodes,))
+
+    barcode_use_dict = {row[0]: row[1] for row in barcode_use_records}
+
+    result = []
+    for bu in barcode_use_list:
+        found = bu['barcode'] in barcode_use_dict
+        result_row = {
+            'barcode': bu['barcode'],
+            'use': bu['use'],
+            'barcode_found': found,
+            'use_match': bu['use']==barcode_use_dict[bu['barcode']] if found else None
```

Comment:
Quick note as I'm skimming: This should casefold the two sides of this comparison, otherwise this defeats the `ci` in `citext` by moving the comparison outside of the database and into Python.

Reply:
Good catch!  I refactored to have the matching occur in the query instead of in Python.